File: src/lib/color.ts
export interface Rgba {
  r: number;
  g: number;
  b: number;
  a: number;
}

const FALLBACK: Rgba = { r: 255, g: 255, b: 255, a: 1 };

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

/**
 * Accepts the CSS colour forms this app can produce: #rgb, #rrggbb, #rrggbbaa,
 * rgb(...) and rgba(...). Anything unparseable falls back to opaque white
 * rather than throwing, because these strings can come from a hand-edited
 * layout.json.
 */
export function parseColor(input: string): Rgba {
  const value = input.trim().toLowerCase();

  if (value.startsWith("#")) {
    const hex = value.slice(1);
    const expand = (c: string) => parseInt(c + c, 16);

    if (hex.length === 3) {
      return { r: expand(hex[0]), g: expand(hex[1]), b: expand(hex[2]), a: 1 };
    }
    if (hex.length === 6 || hex.length === 8) {
      const r = parseInt(hex.slice(0, 2), 16);
      const g = parseInt(hex.slice(2, 4), 16);
      const b = parseInt(hex.slice(4, 6), 16);
      const a = hex.length === 8 ? parseInt(hex.slice(6, 8), 16) / 255 : 1;
      if ([r, g, b].every(Number.isFinite)) return { r, g, b, a };
    }
    return { ...FALLBACK };
  }

  const match = value.match(/^rgba?\(([^)]+)\)$/);
  if (match) {
    const parts = match[1].split(/[,/\s]+/).filter(Boolean).map(Number);
    const [r, g, b, a] = parts;
    if ([r, g, b].every(Number.isFinite)) {
      return {
        r: clamp(Math.round(r), 0, 255),
        g: clamp(Math.round(g), 0, 255),
        b: clamp(Math.round(b), 0, 255),
        a: Number.isFinite(a) ? clamp(a, 0, 1) : 1,
      };
    }
  }

  return { ...FALLBACK };
}

export function toRgbaString({ r, g, b, a }: Rgba): string {
  return `rgba(${r}, ${g}, ${b}, ${Number(a.toFixed(3))})`;
}

/** `#rrggbb`, which is the only form `<input type="color">` understands. */
export function toHex({ r, g, b }: Rgba): string {
  const channel = (c: number) => clamp(Math.round(c), 0, 255).toString(16).padStart(2, "0");
  return `#${channel(r)}${channel(g)}${channel(b)}`;
}

/** Replaces a colour's alpha, preserving its hue. */
export function withAlpha(input: string, alpha: number): string {
  return toRgbaString({ ...parseColor(input), a: clamp(alpha, 0, 1) });
}

/**
 * Solid pressed fill: blends the glow into the key background without dropping
 * alpha (a translucent glow on a transparent overlay window looks invisible).
 */
export function pressedFill(background: string, glow: string, amount = 0.55): string {
  const base = parseColor(background);
  const accent = parseColor(glow);
  const t = clamp(amount, 0, 1);
  return toRgbaString({
    r: Math.round(base.r * (1 - t) + accent.r * t),
    g: Math.round(base.g * (1 - t) + accent.g * t),
    b: Math.round(base.b * (1 - t) + accent.b * t),
    a: Math.max(base.a, 0.9),
  });
}

/** Keeps the alpha from `input` but takes the hue from a `#rrggbb` picker value. */
export function withHex(input: string, hex: string): string {
  const { a } = parseColor(input);
  const { r, g, b } = parseColor(hex);
  return toRgbaString({ r, g, b, a });
}

File: src/color.rs
//! CSS colour parsing used by persisted key styles.

#[derive(Debug, Clone, Copy)]
pub struct Rgba {
    pub r: u8,
    pub g: u8,
    pub b: u8,
    pub a: f32,
}

impl Rgba {
    pub const WHITE: Self = Self {
        r: 255,
        g: 255,
        b: 255,
        a: 1.0,
    };

    pub fn to_egui(self) -> egui::Color32 {
        egui::Color32::from_rgba_unmultiplied(
            self.r,
            self.g,
            self.b,
            (self.a.clamp(0.0, 1.0) * 255.0).round() as u8,
        )
    }

    pub fn to_css(self) -> String {
        format!(
            "rgba({}, {}, {}, {})",
            self.r,
            self.g,
            self.b,
            (self.a * 1000.0).round() / 1000.0
        )
    }
}

fn clamp01(v: f32) -> f32 {
    v.clamp(0.0, 1.0)
}

pub fn parse_color(input: &str) -> Rgba {
    let value = input.trim().to_lowercase();
    if let Some(hex) = value.strip_prefix('#') {
        return parse_hex(hex).unwrap_or(Rgba::WHITE);
    }
    if let Some(inner) = value
        .strip_prefix("rgba(")
        .or_else(|| value.strip_prefix("rgb("))
        .and_then(|s| s.strip_suffix(')'))
    {
        let parts: Vec<f32> = inner
            .split(|c: char| c == ',' || c == '/' || c.is_whitespace())
            .filter(|p| !p.is_empty())
            .filter_map(|p| p.parse().ok())
            .collect();
        if parts.len() >= 3 {
            return Rgba {
                r: parts[0].round().clamp(0.0, 255.0) as u8,
                g: parts[1].round().clamp(0.0, 255.0) as u8,
                b: parts[2].round().clamp(0.0, 255.0) as u8,
                a: parts.get(3).copied().map(clamp01).unwrap_or(1.0),
            };
        }
    }
    Rgba::WHITE
}

fn parse_hex(hex: &str) -> Option<Rgba> {
    let expand = |c: u8| -> u8 { let v = c; (v << 4) | v };
    match hex.len() {
        3 => {
            let bytes = hex.as_bytes();
            Some(Rgba {
                r: expand(from_hex(bytes[0])?),
                g: expand(from_hex(bytes[1])?),
                b: expand(from_hex(bytes[2])?),
                a: 1.0,
            })
        }
        6 | 8 => {
            let r = u8::from_str_radix(&hex[0..2], 16).ok()?;
            let g = u8::from_str_radix(&hex[2..4], 16).ok()?;
            let b = u8::from_str_radix(&hex[4..6], 16).ok()?;
            let a = if hex.len() == 8 {
                u8::from_str_radix(&hex[6..8], 16).ok()? as f32 / 255.0
            } else {
                1.0
            };
            Some(Rgba { r, g, b, a })
        }
        _ => None,
    }
}

fn from_hex(c: u8) -> Option<u8> {
    match c {
        b'0'..=b'9' => Some(c - b'0'),
        b'a'..=b'f' => Some(c - b'a' + 10),
        _ => None,
    }
}

/// Opaque pressed fill: blend glow into background without dropping alpha.
pub fn pressed_fill(background: &str, glow: &str, amount: f32) -> Rgba {
    let base = parse_color(background);
    let accent = parse_color(glow);
    let t = clamp01(amount);
    Rgba {
        r: ((base.r as f32) * (1.0 - t) + (accent.r as f32) * t).round() as u8,
        g: ((base.g as f32) * (1.0 - t) + (accent.g as f32) * t).round() as u8,
        b: ((base.b as f32) * (1.0 - t) + (accent.b as f32) * t).round() as u8,
        a: base.a.max(0.9),
    }
}

pub fn to_hex(input: &str) -> String {
    let c = parse_color(input);
    format!("#{:02x}{:02x}{:02x}", c.r, c.g, c.b)
}

pub fn with_hex(input: &str, hex: &str) -> String {
    let a = parse_color(input).a;
    let mut c = parse_color(hex);
    c.a = a;
    c.to_css()
}
